[Mono] Arrays shouldn't implement linked out generic classes

lambdageek opened the pull request:
Don't implement the `IList<T>` and `IReadOnlyList<T>` for array types if those interfaces are linked out.
Also check for individual methods from the generic interfaces - if the methods aren't used and are linked out, drop them from the array initialization.

Also drop option for pre-.NET 2.0 behavior (in netcore and in mono net 4.x mode)

Related to #44859

lambdageek committed:
[runtime] IList<T> and IReadOnlyList<T> can be linked out

Allow mono.defaults.generic_ilist_class to be null.

When creating arrays, if the interfaces aren't present, don't implement them.

lambdageek committed:
[class] Don't implement linked out colleciton methods on arrays

If the method doesn't exist on the generic class, don't implement it in the
array subclass.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -4610,7 +4610,16 @@ method_nonpublic (MonoMethod* method, gboolean start_klass)
+			 */
+			return TRUE;
+#else
 			return (start_klass || mono_defaults.generic_ilist_class);
```
We don't support 1.0 anymore so this probably could be removed.

vargaz commented on src/mono/mono/metadata/icall.c:
```diff
@@ -5153,11 +5162,18 @@ ves_icall_System_Reflection_Assembly_InternalGetType (MonoReflectionAssemblyHand
 			/* 1.0 and 2.0 throw different exceptions */
 			if (mono_defaults.generic_ilist_class)
-				mono_error_set_argument (error, NULL, "Type names passed to Assembly.GetType() must not specify an assembly.");
+				set_argument_exn = TRUE;
```
Ditto.

lambdageek committed:
[mono] Drop pre-.NET 2.0 behavior

in method_nonpublic, assembly visibility is always non-public.

in InternalGetType, always throw an ArgumentException (not a TypeLoadException)
if a type name includes an assembly name.

lambdageek committed:
fix typo

lambdageek committed:
fixup return the correct generic array method count

lambdageek committed:
Update linker descriptor not to keep the generic collections

lambdageek commented:
Verified that with this PR, `IReadOnlyCollection<T>.get_Count` gets trimmed away.

There's some followup work that we can do once https://github.com/mono/linker/issues/1636 is resolved: we will be able to remove the `InternalArray__` methods from the linker descriptor file and only keep the implementations if the relevant interfaces are preserved.
